File: src/main/kotlin/MapWayFinder.java
import model.Game;
import model.LaneType;
import model.Wizard;
import model.World;

import java.util.*;
import java.util.stream.Collectors;

public class MapWayFinder {

    private World world;
    private Game game;
    private Wizard wizard;
    private MapHelper mapHelper;

    public static final double NEXT_LINE_DISTANCE = 300D;
    public static final double WAY_FINDER_DISTANCE = 250D;

    public static final double NEXT_LINE_DISTANCE_MULTIPLIER = 1.1;

    public MapWayFinder(World world, Game game, Wizard wizard) {
        this.wizard = wizard;
        this.mapHelper = new MapHelper(world, game, wizard);
    }

    public Point2D getNextWaypoint(LaneType laneType) {
        List<LinePosition> wizardOnLine = mapHelper.getLinePositions(wizard, 1).stream()
                .filter(linePosition -> linePosition.getMapLine().getLaneType() == laneType)
                .collect(Collectors.toList());

        Point2D linePointToBaseEnemy = mapHelper.getLinePointToBaseEnemy(laneType);
        if (wizardOnLine.size() > 0) {
            return getPointTo(linePointToBaseEnemy, false);
        } else
            return getPointTo(linePointToBaseEnemy, true);

    }

    public Point2D getPreviousWaypoint(LaneType laneType) {
        return getPointTo(mapHelper.friendBasePoint, true);
    }

    public Point2D getPointTo(Point2D point, boolean safeWay) {
        List<LinePosition> wizardPositions = mapHelper.getLinePositions(wizard, 1);

        double pointDistance = point.getDistanceTo(wizard);

        if (pointDistance < WAY_FINDER_DISTANCE)
            return point;

        if (wizardPositions == null || wizardPositions.size() == 0) {
            return mapHelper.getNearestPointInLine(wizard);
        } else {
            List<LinePosition> linePositions = mapHelper.getLinePositions(point.getX(), point.getY(), 1);

            for (LinePosition wizardPosition : wizardPositions) {
                for (LinePosition linePosition : linePositions) {
                    if (linePositions.isEmpty())
                        throw new RuntimeException("asfd");

                    MapLine wizardLine = wizardPosition.getMapLine();
                    if (wizardLine.equals(linePosition.getMapLine())) {
                        if (safeWay) {
                            if (wizardLine.getEnemyWizardPositions().isEmpty())
                                return mapHelper.getPointInLine(linePosition);

                            boolean isSafeWay = wizardLine.getEnemyWizardPositions().values().stream()
                                    .noneMatch(value -> isInRange(value, wizardPosition.getPosition(), linePosition.getPosition()));

                            if (isSafeWay)
                                return mapHelper.getPointInLine(linePosition);
                        } else
                            return mapHelper.getPointInLine(linePosition);
                    }
                }
            }

            Optional<Pair<Double, Point2D>> findWayPoint = wizardPositions.stream()
                    .map(wizardPosition -> findMapWayPoint(wizardPosition, linePositions, safeWay))
                    .filter(Optional::isPresent)
                    .map(Optional::get)
                    .min(Comparator.comparing(Pair::getFirst));

            if (!findWayPoint.isPresent() && safeWay)
                return wizardPositions.stream()
                        .map(wizardPosition -> findMapWayPoint(wizardPosition, linePositions, false))
                        .filter(Optional::isPresent)
                        .map(Optional::get)
                        .min(Comparator.comparing(Pair::getFirst)).get().getSecond();

            return findWayPoint.get().getSecond();
        }
    }

    private boolean isInRange(double checkedValue, double start, double end) {
        double minValue = StrictMath.min(start, end);
        double maxValue = StrictMath.max(start, end);

        if (checkedValue < minValue || checkedValue > maxValue) return false;
        return true;
    }

    private Optional<Pair<Double, Point2D>> findMapWayPoint(LinePosition wizardLinePosition, List<LinePosition> pointLinePositions, boolean safeWay) {
        MapLine wizardMapLine = wizardLinePosition.getMapLine();

        final List<MapLine> startLines = wizardMapLine.getStartLines();
        final List<MapLine> endLines = wizardMapLine.getEndLines();

        if (safeWay) {
            Map<Point2D, Double> enemyWizardPositions = wizardMapLine.getEnemyWizardPositions();

            if (enemyWizardPositions.size() > 0) {
                boolean isStartSafe = enemyWizardPositions.values().stream()
                        .allMatch(enemyWizardsLocation -> enemyWizardsLocation >= wizardLinePosition.getPosition());

                if (!isStartSafe)
                    startLines.clear();

                boolean isEndSafe = enemyWizardPositions.values().stream()
                        .allMatch(enemyWizardsLocation -> enemyWizardsLocation <= wizardLinePosition.getPosition());

                if (!isEndSafe)
                    endLines.clear();
            }
        }

        List<MapLine> allLines = new ArrayList<>();
        allLines.addAll(startLines);
        allLines.addAll(endLines);

        List<MapLine> searchingLines = pointLinePositions.stream().map(LinePosition::getMapLine)
                .collect(Collectors.toList());


        List<Pair<Double, MapLine>> findWays = allLines.stream()
                .filter(searchingLines::contains)
                .map(mapLine -> {
                    LinePosition pointPosition = pointLinePositions.stream()
                            .filter(linePosition -> linePosition.getMapLine().equals(mapLine))
                            .findFirst().get();

                    return new Pair<>(mapLine, pointPosition);
                })
                .filter(mapLine -> {
                    if (!safeWay) return true;

                    if (mapLine.getFirst().getMapLineStatus() == MapLineStatus.GREEN) return true;

                    if (mapLine.getFirst().getStartPoint().equals(wizardMapLine.getStartPoint())
                            || mapLine.getFirst().getStartPoint().equals(wizardMapLine.getEndPoint()))
                        return mapLine.getFirst().getEnemyWizardPositions().values().stream()
                                .anyMatch(position -> mapLine.getSecond().getPosition() < position);
                    else
                        return mapLine.getFirst().getEnemyWizardPositions().values().stream()
                                .anyMatch(position -> mapLine.getSecond().getPosition() > position);
                })
                .map(mapLine -> {
                    boolean ifStartLine = startLines.contains(mapLine.getFirst());
                    double startDistance = wizardLinePosition.getPosition();
                    if (!ifStartLine)
                        startDistance = wizardMapLine.getLineLength() - wizardLinePosition.getPosition();

                    if (mapLine.getFirst().getStartPoint().equals(wizardMapLine.getStartPoint())
                            || mapLine.getFirst().getStartPoint().equals(wizardMapLine.getEndPoint()))
                        return new Pair<>(startDistance + mapLine.getSecond().getPosition(), mapLine.getFirst());
                    else
                        return new Pair<>(startDistance + (mapLine.getFirst().getLineLength() - mapLine.getSecond().getPosition()),
                                mapLine.getFirst());
                })
                .collect(Collectors.toList());

        if (findWays.isEmpty()) {
            if (startLines != null) {
                Point2D startPoint = wizardMapLine.getStartPoint();
                double startPosition = wizardLinePosition.getPosition();

                List<Pair<Double, MapLine>> startWays = getWays(
                        pointLinePositions, safeWay, startLines, startPoint, startPosition, wizardMapLine);

                findWays.addAll(startWays);
            }

            if (endLines != null) {
                List<Pair<Double, MapLine>> stopWays = getWays(
                        pointLinePositions, safeWay, endLines, wizardMapLine.getEndPoint(),
                        wizardMapLine.getLineLength() - wizardLinePosition.getPosition(), wizardMapLine);

                findWays.addAll(stopWays);
            }
        }


        Optional<Pair<Double, Point2D>> minValue = findWays.stream()
                .min(Comparator.comparing(Pair::getFirst))
                .map(pair -> new Pair<>(pair.getFirst(), getPointFromMapLine(wizardLinePosition, pair.getSecond())));

        return minValue;
    }

    private Point2D getPointFromMapLine(LinePosition wizardLinePosition, MapLine line) {
        Double positionToLinePoint = wizardLinePosition.getPosition();
        Point2D wayPoint = wizardLinePosition.getMapLine().getStartPoint();
        if (!wizardLinePosition.getMapLine().getStartLines().contains(line)) {
            positionToLinePoint = wizardLinePosition.getMapLine().getLineLength() - positionToLinePoint;
            wayPoint = wizardLinePosition.getMapLine().getEndPoint();
        }

        if (positionToLinePoint >= NEXT_LINE_DISTANCE) {
            return wayPoint;
        } else {
            double linePosition = NEXT_LINE_DISTANCE * NEXT_LINE_DISTANCE_MULTIPLIER;
            if (line.getEndPoint().equals(wayPoint))
                linePosition = line.getLineLength() - linePosition;

            return mapHelper.getPointInLine(new LinePosition(line, linePosition));
        }
    }


    private List<Pair<Double, MapLine>> getWays(List<LinePosition> pointLinePositions, boolean safeWay, List<MapLine> startLines, Point2D startPoint, double startPosition, MapLine wizardLine) {
        return startLines.stream()
                .map(mapLine -> {
                    List<MapLine> checkedLines = new ArrayList<>();
                    checkedLines.add(wizardLine);

                    WayParams wayParams = new WayParams(
                            mapLine,
                            mapLine,
                            startPoint,
                            startPosition,
                            pointLinePositions,
                            safeWay,
                            checkedLines
                    );

                    return findMapWayPoint(wayParams);
                })
                .filter(Optional::isPresent)
                .map(Optional::get)
                .collect(Collectors.toList());
    }

    private Optional<Pair<Double, MapLine>> findMapWayPoint(WayParams wayParams) {
        if (wayParams.getCheckedLines().size() > 3) return Optional.empty();

        MapLine mapLine = wayParams.getMapLine();

        List<MapLine> furtherLines = mapLine.getStartLines();
        if (wayParams.getStartPoint().equals(mapLine.getEndPoint()))
            furtherLines = mapLine.getEndLines();

        List<MapLine> filteredFurtherLines = furtherLines.stream()
                .filter(furtherLine -> !wayParams.getCheckedLines().contains(furtherLine))
                .collect(Collectors.toList());

        if (filteredFurtherLines.isEmpty()) return Optional.empty();

        Optional<LinePosition> findWay = wayParams.getPointLinePositions().stream()
                .filter(linePosition -> {
                    MapLine checkedMapLine = linePosition.getMapLine();
                    if (filteredFurtherLines.contains(checkedMapLine)) {
                        if (!wayParams.isSafeWay()) return true;

                        if (checkedMapLine.getMapLineStatus() == MapLineStatus.GREEN) return true;

                        if (checkedMapLine.getStartPoint().equals(wayParams.getStartPoint()))
                            return checkedMapLine.getEnemyWizardPositions().values().stream()
                                    .anyMatch(position -> linePosition.getPosition() < position);
                        else
                            return checkedMapLine.getEnemyWizardPositions().values().stream()
                                    .anyMatch(position -> linePosition.getPosition() > position);

                    } else
                        return false;
                })
                .findFirst();

        if (findWay.isPresent()) {
            double newLength = wayParams.getStartDestination();
            if (wayParams.getStartPoint().equals(findWay.get().getMapLine().getStartPoint()))
                newLength += findWay.get().getPosition();
            else
                newLength += findWay.get().getMapLine().getLineLength() - findWay.get().getPosition();

            return Optional.of(new Pair<>(newLength, wayParams.getStartMapLine()));
        } else {
            return filteredFurtherLines.stream()
                    .map(furtherLine -> {
                        List<MapLine> newWayLines = new ArrayList<>();
                        newWayLines.addAll(wayParams.getCheckedLines());
                        newWayLines.add(furtherLine);

                        Point2D nextPoint = furtherLine.getStartPoint();
                        if (!furtherLine.getStartPoint().equals(wayParams.getMapLine().getEndPoint()))
                            nextPoint = furtherLine.getEndPoint();

                        WayParams newWayParams = new WayParams(
                                wayParams.getStartMapLine(),
                                furtherLine,
                                nextPoint,
                                wayParams.getStartDestination() + wayParams.getMapLine().getLineLength(),
                                wayParams.getPointLinePositions(),
                                wayParams.isSafeWay(),
                                newWayLines
                        );

                        return findMapWayPoint(newWayParams);
                    })
                    .filter(Optional::isPresent)
                    .map(Optional::get)
                    .min(Comparator.comparing(Pair::getFirst));
        }

    }


}

File: src/main/kotlin/MapWayFinder.kt
import MapHelper.friendBasePoint
import MapHelper.getLinePointToBaseEnemy
import MapHelper.getLinePositions
import MapHelper.getNearestPointInLine
import MapHelper.getPointInLine
import model.Game
import model.LaneType
import model.Wizard
import model.World
import java.util.*

class MapWayFinder(val world: World, val game: Game, private val wizard: Wizard) {

    fun getNextWaypoint(laneType: LaneType, globalStrateg: GlobalStrateg): Point2D {
        val wizardOnLine = getLinePositions(wizard, 1.0)
                .filter { linePosition -> linePosition.mapLine.laneType === laneType }

        val linePointToBaseEnemy = getLinePointToBaseEnemy(laneType, globalStrateg)

        val safeWay = wizardOnLine.isEmpty() || wizard.getDistanceTo(linePointToBaseEnemy) > MAX_SAFE_DISTANCE
                || globalStrateg == GlobalStrateg.DEFENCE

        val pointToMove = getPointTo(linePointToBaseEnemy, safeWay)

        return if (wizard.getDistanceTo(pointToMove) > MAX_SELF_WAY_RANGE) {
            getPointTo(correctPoint(pointToMove), safeWay)
        } else
            pointToMove
    }

    private fun correctPoint(point: Point2D): Point2D {
        if (world.tickIndex - lastNextCheck >= MIN_TICK_DIFF_TO_FIND_POINT || previousNextPoint == null) {
            lastNextCheck = world.tickIndex
            previousNextPoint = point
        }

        return previousNextPoint!!
    }

    fun getPreviousWaypoint(laneType: LaneType): Point2D {
        return if (wizard.getDistanceTo(friendBasePoint) > MAX_SELF_WAY_RANGE) {
            getPointTo(friendBasePoint, true)
        } else
            friendBasePoint

    }

    fun getPointTo(point: Point2D, safeWay: Boolean): Point2D {
        if (!safeWay)
            return getSafePointTo(point, safeWay)!!

        return getSafePointTo(point, safeWay) ?: getSafePointTo(point, false)!!
    }

    fun getSafePointTo(point: Point2D, safeWay: Boolean): Point2D? {
        val wizardPositions = getLinePositions(wizard.toPoint(), 1.0)

        val pointDistance = wizard.getDistanceTo(point)

        if (pointDistance < WAY_FINDER_DISTANCE)
            return point

        if (wizardPositions.isEmpty()) {
            return getNearestPointInLine(wizard.toPoint())
        } else {
            val linePositions = getLinePositions(point.x, point.y, 1.0)

            for ((wizardLine, position) in wizardPositions) {
                for (linePosition in linePositions) {
                    if (linePositions.isEmpty())
                        throw RuntimeException("how??? how?????")

                    if (wizardLine == linePosition.mapLine) {
                        if (safeWay) {
                            if (wizardLine.enemyWizardPositions.isEmpty())
                                return getPointInLine(linePosition)

                            if (wizardLine.startPoint == friendBasePoint && wizardLine.enemyWizardPositions
                                    .all { it.value < wizardLine.lineLength * BASE_POINT_POSITION_FACTOR })
                                return getPointInLine(linePosition)

                            val isSafeWay = wizardLine.enemyWizardPositions.values
                                    .none { value -> isInRange(value, position, linePosition.position) }

                            if (isSafeWay)
                                return getPointInLine(linePosition)
                        } else
                            return getPointInLine(linePosition)
                    }
                }
            }

            return wizardPositions
                    .mapNotNull { wizardPosition -> findMapWayPoint(wizardPosition, linePositions, safeWay) }
                    .minBy { it.first }?.second
        }
    }

    private fun isInRange(checkedValue: Double, start: Double, end: Double): Boolean {
        val minValue = StrictMath.min(start, end)
        val maxValue = StrictMath.max(start, end)

        if (checkedValue < minValue || checkedValue > maxValue) return false
        return true
    }

    private fun findMapWayPoint(wizardLinePosition: LinePosition, pointLinePositions: List<LinePosition>, safeWay: Boolean): Pair<Double, Point2D>? {
        val wizardMapLine = wizardLinePosition.mapLine

        val checkWays = listOf(
                wizardMapLine.startPoint to wizardLinePosition.position,
                wizardMapLine.endPoint to (wizardMapLine.lineLength - wizardLinePosition.position)
        )

        return checkWays.mapNotNull { checkWay ->
            val wayParam = WayParams(
                    startMapLine = wizardMapLine,
                    checkedLines = listOf(wizardMapLine),
                    isSafeWay = safeWay,
                    pointLinePositions = pointLinePositions,
                    startDestination = checkWay.second,
                    startPoint = checkWay.first
            )

            findMapWayPoint(wayParam)
        }.minBy { it.first }?.let { value ->
            value.first to getPointFromMapLine(wizardLinePosition, value.second)
        }
    }

    private fun getPointFromMapLine(wizardLinePosition: LinePosition, line: MapLine): Point2D {
        var positionToLinePoint: Double? = wizardLinePosition.position
        var wayPoint = wizardLinePosition.mapLine.startPoint
        if (!wizardLinePosition.mapLine.startLines.contains(line)) {
            positionToLinePoint = wizardLinePosition.mapLine.lineLength - positionToLinePoint!!
            wayPoint = wizardLinePosition.mapLine.endPoint
        }

        if ((positionToLinePoint ?: 0.0) >= NEXT_LINE_DISTANCE) {
            return wayPoint
        } else {
            var linePosition = NEXT_LINE_DISTANCE * NEXT_LINE_DISTANCE_MULTIPLIER
            if (line.endPoint == wayPoint)
                linePosition = line.lineLength - linePosition

            return getPointInLine(LinePosition(line, linePosition))
        }
    }


    private fun findMapWayPoint(wayParams: WayParams): Pair<Double, MapLine>? {
        if (wayParams.checkedLines.size > 3) return null

        val mapLine = wayParams.startMapLine

        var furtherLines = mapLine.startLines
        if (wayParams.startPoint == mapLine.endPoint)
            furtherLines = mapLine.endLines

        val filteredFurtherLines: List<MapLine> = furtherLines
                .filter { furtherLine -> !wayParams.checkedLines.contains(furtherLine) }

        if (filteredFurtherLines.isEmpty()) return null

        val findWay: LinePosition? = wayParams.pointLinePositions
                .filter { linePosition ->
                    val checkedMapLine = linePosition.mapLine
                    if (filteredFurtherLines.contains(checkedMapLine)) {
                        if (!wayParams.isSafeWay) return@filter true

                        if (checkedMapLine.startPoint == wayParams.startPoint)
                            return@filter checkedMapLine.enemyWizardPositions.values
                                    .all { position ->
                                        linePosition.position < position
                                    } || ((checkedMapLine.startPoint == friendBasePoint) && checkedMapLine.enemyWizardPositions.values
                                    .all { position ->
                                        linePosition.position < position || position < checkedMapLine.lineLength * BASE_POINT_POSITION_FACTOR
                                    })
                        else
                            return@filter checkedMapLine.enemyWizardPositions.values
                                    .all { position ->
                                        linePosition.position > position
                                    } || ((checkedMapLine.startPoint == friendBasePoint) && checkedMapLine.enemyWizardPositions.values
                                    .all { position ->
                                        linePosition.position > position || position < checkedMapLine.lineLength * BASE_POINT_POSITION_FACTOR
                                    })
                    } else
                        false
                }.firstOrNull()

        if (findWay != null) {
            var newLength = wayParams.startDestination!!
            if (wayParams.startPoint == findWay.mapLine.startPoint)
                newLength += findWay.position
            else
                newLength += findWay.mapLine.lineLength - findWay.position

            return newLength to (wayParams.mapLine ?: findWay.mapLine)
        } else {
            return filteredFurtherLines
                    .mapNotNull { furtherLine ->
                        val newWayLines = ArrayList<MapLine>()
                        newWayLines.addAll(wayParams.checkedLines)
                        newWayLines.add(furtherLine)

                        var nextPoint = furtherLine.startPoint
                        if (furtherLine.startPoint != wayParams.startMapLine.endPoint)
                            nextPoint = furtherLine.endPoint

                        val newWayParams = WayParams(
                                startMapLine = furtherLine,
                                startPoint = nextPoint,
                                startDestination = wayParams.startDestination!! + furtherLine.lineLength,
                                pointLinePositions = wayParams.pointLinePositions,
                                mapLine = wayParams.mapLine ?: furtherLine,
                                isSafeWay = wayParams.isSafeWay,
                                checkedLines = newWayLines
                        )

                        findMapWayPoint(newWayParams)
                    }.minBy { it.first }
        }
    }

    companion object {

        var previousNextPoint: Point2D? = null
        var lastNextCheck: Int = 0

        val MAX_SELF_WAY_RANGE = 150

        val MIN_TICK_DIFF_TO_FIND_POINT = 50

        val NEXT_LINE_DISTANCE: Double = 450.0

        val WAY_FINDER_DISTANCE = 150.0

        val NEXT_LINE_DISTANCE_MULTIPLIER = 1.1

        val BASE_POINT_POSITION_FACTOR: Double = 0.3

        val MAX_SAFE_DISTANCE: Double = 1000.0
    }


}
